flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/AlternatingController.java

```diff
@@ -146,28 +146,7 @@ private void switchToUnaligned(
 
 	@Override
 	public Optional<CheckpointBarrier> postProcessLastBarrier(InputChannelInfo channelInfo, CheckpointBarrier barrier) throws IOException, CheckpointException {
-		Optional<CheckpointBarrier> maybeTimeOut = asTimedOut(barrier);
-		if (maybeTimeOut.isPresent() && activeController == alignedController) {
-			switchToUnaligned(channelInfo, maybeTimeOut.get());
-			checkState(activeController == unalignedController);
-			checkState(!activeController.postProcessLastBarrier(channelInfo, maybeTimeOut.orElse(barrier)).isPresent());
-			return maybeTimeOut;
-		}
-
-		barrier = maybeTimeOut.orElse(barrier);
-		if (barrier.getCheckpointOptions().isUnalignedCheckpoint()) {
-			checkState(activeController == unalignedController);
-			checkState(!activeController.postProcessLastBarrier(channelInfo, maybeTimeOut.orElse(barrier)).isPresent());
-			return Optional.empty();
-		}
-		else {
-			checkState(activeController == alignedController);
-			Optional<CheckpointBarrier> triggerResult = activeController.postProcessLastBarrier(
-				channelInfo,
-				barrier);
-			checkState(triggerResult.isPresent());
-			return triggerResult;
-		}
```

Comment:
Why have you removed this code? What was the problem?

Was it subsumed by `switchToUnaligned ` call happening in the last `barrierReceived` call?

Reply:
Yes, with `switchToUnaligned` in `barrierReceived` it seems unnecessary.
With this code, `UnalignedCheckpointITCase` fails: some tests on `checkState` (line 164 above), some hanging up.

Besides, why timeout alignment if it's the last barrier? This essentially means that alignment is done.

Reply:
> Besides, why timeout alignment if it's the last barrier? This essentially means that alignment is done.

The alignment is done, but that's just input. As we do not have code to timeout outputs, it's better to timeout to UC even if the alignment was completed (although too late). Think especially about a case with just a single input channel.

Reply:
Isn't it enough to just convert the barrier upon triggering checkpoint in subtask?

This is what I've done in `[FLINK-19681][checkpointing] Use converted barrier after disabling alignment`. It is needed anyways for correct handling in `SubtaskCheckpointCoordinatorImpl`.